File: macos-app/AgentServerSwiftTests/Sources/AgentServerCore/RunReview.swift
import Foundation

/// A consumer-facing account of one local assistant run.
public struct RunReview: Decodable, Equatable, Sendable {
    public let outcome: RunReviewOutcome
    public let headline: PresentationStatement
    public let summary: PresentationStatement
    public let accomplishments: [PresentationStatement]
    public let changes: [PresentationStatement]
    public let outputs: [PresentationStatement]
    public let problems: [PresentationStatement]
    public let suggestions: [PresentationStatement]
    public let timeline: [HumanTimelineEntry]
    public let operationalCompleteness: OperationalCompleteness
    let waiting: RunReviewWaiting?
    public let technicalDetailsReference: String

    enum CodingKeys: String, CodingKey {
        case outcome, headline, summary, accomplishments, changes, outputs, problems
        case suggestions, timeline, operationalCompleteness, waiting, technicalDetailsReference
    }

    init(
        outcome: RunReviewOutcome,
        headline: PresentationStatement,
        summary: PresentationStatement,
        accomplishments: [PresentationStatement],
        changes: [PresentationStatement],
        outputs: [PresentationStatement],
        problems: [PresentationStatement],
        suggestions: [PresentationStatement],
        timeline: [HumanTimelineEntry],
        operationalCompleteness: OperationalCompleteness,
        waiting: RunReviewWaiting? = nil,
        technicalDetailsReference: String
    ) {
        self.outcome = outcome
        self.headline = headline
        self.summary = summary
        self.accomplishments = accomplishments
        self.changes = changes
        self.outputs = outputs
        self.problems = problems
        self.suggestions = suggestions
        self.timeline = timeline
        self.operationalCompleteness = operationalCompleteness
        self.waiting = waiting
        self.technicalDetailsReference = technicalDetailsReference
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        outcome = try container.decode(RunReviewOutcome.self, forKey: .outcome)
        headline = try container.decode(PresentationStatement.self, forKey: .headline)
        summary = try container.decode(PresentationStatement.self, forKey: .summary)
        accomplishments = try container.decode([PresentationStatement].self, forKey: .accomplishments)
        changes = try container.decode([PresentationStatement].self, forKey: .changes)
        outputs = try container.decode([PresentationStatement].self, forKey: .outputs)
        problems = try container.decode([PresentationStatement].self, forKey: .problems)
        suggestions = try container.decode([PresentationStatement].self, forKey: .suggestions)
        timeline = try container.decode([HumanTimelineEntry].self, forKey: .timeline)
        operationalCompleteness = try container.decode(
            OperationalCompleteness.self,
            forKey: .operationalCompleteness
        )
        waiting = try container.decodeIfPresent(RunReviewWaiting.self, forKey: .waiting)
        technicalDetailsReference = try container.decode(
            String.self,
            forKey: .technicalDetailsReference
        )
    }
}

struct RunReviewWaiting: Decodable, Equatable, Sendable {
    let waitingFor: PresentationStatement
    let reason: PresentationStatement
    let userAction: PresentationAction?
    let expiresAt: Date?

    private enum CodingKeys: String, CodingKey {
        case waitingFor, reason, userAction, expiresAt
    }

    init(
        waitingFor: PresentationStatement,
        reason: PresentationStatement,
        userAction: PresentationAction?,
        expiresAt: Date?
    ) {
        self.waitingFor = waitingFor
        self.reason = reason
        self.userAction = userAction
        self.expiresAt = expiresAt
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        waitingFor = try container.decode(PresentationStatement.self, forKey: .waitingFor)
        reason = try container.decode(PresentationStatement.self, forKey: .reason)
        userAction = try container.decodeIfPresent(PresentationAction.self, forKey: .userAction)
        guard let value = try container.decodeIfPresent(String.self, forKey: .expiresAt) else {
            expiresAt = nil
            return
        }
        guard let date = RunReviewDateParser.date(from: value) else {
            throw DecodingError.dataCorruptedError(
                forKey: .expiresAt,
                in: container,
                debugDescription: "Expected an ISO 8601 date."
            )
        }
        expiresAt = date
    }
}

private enum RunReviewDateParser {
    static func date(from value: String) -> Date? {
        let fractional = ISO8601DateFormatter()
        fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return fractional.date(from: value) ?? ISO8601DateFormatter().date(from: value)
    }
}

/// The user-visible result of a run.
public enum RunReviewOutcome: String, Decodable, Equatable, Sendable {
    case succeeded
    case partial
    case failed
    case canceled
    case skipped
    case working
    case waiting
    case unknown
}

/// Consumer copy paired with the local evidence that supports it.
public struct PresentationStatement: Decodable, Equatable, Sendable {
    public let text: String
    public let evidenceReferences: [String]
}

/// A meaningful step in an assistant run.
public struct HumanTimelineEntry: Decodable, Equatable, Sendable {
    public let kind: HumanTimelineEntryKind
    public let label: PresentationStatement
    public let occurredAt: String?
}

/// The stable vocabulary used by the human timeline.
public enum HumanTimelineEntryKind: String, Decodable, Equatable, Sendable {
    case started
    case connected
    case read
    case changed
    case produced
    case waiting
    case resumed
    case problem
    case finished
}

/// Whether deterministic output requirements were met.
public enum OperationalCompleteness: String, Decodable, Equatable, Sendable {
    case complete
    case incomplete
    case notAssessed = "not_assessed"
}

File: server-app/src/presentation/run-review.ts
import { basename } from 'node:path';
import type { PendingInteraction } from '../interaction/store.js';
import type { StoredRun } from '../reporting/store.js';
import {
  RunReviewSchema,
  type HumanTimelineEntry,
  type PresentationStatement,
  type RunReview,
  type RunReviewWaiting,
} from './models.js';

type RequiredOutput = {
  label: string;
};

export type CreateRunReviewInput = {
  run: StoredRun;
  requiredOutput?: RequiredOutput;
  observedAccomplishments?: PresentationStatement[];
  pendingInteraction?: PendingInteraction;
  now?: Date;
};

type ReviewMeaning = Pick<
  RunReview,
  'outcome' | 'headline' | 'summary' | 'problems' | 'suggestions' | 'operationalCompleteness'
>;

function statement(text: string, ...evidenceReferences: string[]): PresentationStatement {
  return { text, evidenceReferences };
}

function safeFileLabel(path: string): string {
  const label = basename(path.trim());
  return label || 'a local file';
}

function currentPendingInteraction(
  run: StoredRun,
  interaction: PendingInteraction | undefined,
  now: Date,
): PendingInteraction | undefined {
  if (
    run.status !== 'running'
    || interaction?.status !== 'pending'
    || interaction.runId !== run.runId
    || interaction.agentId !== run.agentId
    || interaction.expiresAt <= now
  ) {
    return undefined;
  }
  return interaction;
}

function waitingActionLabel(interaction: PendingInteraction): string {
  if ((interaction.request.options?.length ?? 0) > 0) return 'Choose';
  if (interaction.request.freeText) return 'Answer';
  return 'Review';
}

function waitingPresentation(interaction: PendingInteraction): RunReviewWaiting {
  return {
    waitingFor: statement(
      interaction.request.message,
      'interaction.request.message',
    ),
    reason: statement(
      'The assistant needs your response before it can continue.',
      'interaction.status',
      'interaction.runId',
    ),
    userAction: {
      kind: 'respond',
      label: waitingActionLabel(interaction),
      targetReference: `interaction:${interaction.id}`,
    },
    expiresAt: interaction.expiresAt.toISOString(),
  };
}

function completedMeaning(
  run: StoredRun,
  requiredOutput: RequiredOutput | undefined,
): ReviewMeaning {
  return {
    outcome: 'succeeded',
    headline: statement(`${run.agentName} finished`, 'run.status'),
    summary: run.summary
      ? statement(run.summary, 'run.summary')
      : statement(
        'The run finished, but it did not provide a summary.',
        'run.status',
        'run.summary',
      ),
    problems: [],
    suggestions: [],
    operationalCompleteness: requiredOutput ? 'complete' : 'not_assessed',
  };
}

function failedMeaning(
  run: StoredRun,
  requiredOutput: RequiredOutput | undefined,
): ReviewMeaning {
  if (run.code === 'user_canceled' || run.code === 'run_canceled') {
    return {
      outcome: 'canceled',
      headline: statement(`${run.agentName} was canceled`, 'run.code'),
      summary: statement(
        'The run was canceled. No further work was allowed.',
        'run.code',
      ),
      problems: [],
      suggestions: [],
      operationalCompleteness: requiredOutput ? 'incomplete' : 'not_assessed',
    };
  }

  if (run.code === 'output_contract_unmet') {
    const label = requiredOutput?.label ?? 'output';
    return {
      outcome: 'partial',
      headline: statement(`${run.agentName} did not finish`, 'run.status', 'run.code'),
      summary: statement(
        `The run stopped because the required ${label} was not produced.`,
        'run.code',
        ...(requiredOutput ? ['agent.output.primary'] : []),
      ),
      problems: [statement(
        `The required ${label} was not produced.`,
        'run.code',
        ...(requiredOutput ? ['agent.output.primary'] : []),
      )],
      suggestions: [statement('Review the output settings before retrying.', 'run.code')],
      operationalCompleteness: 'incomplete',
    };
  }

  if (run.error && /\b(?:spawn\s+\S+\s+enoent|executable not found)\b/i.test(run.error)) {
    return {
      outcome: 'failed',
      headline: statement(`${run.agentName} could not start`, 'run.status', 'run.error'),
      summary: statement(
        'The AI engine could not start. Check that it is installed and signed in.',
        'run.error',
      ),
      problems: [statement('The AI engine is unavailable.', 'run.error')],
      suggestions: [statement('Open Connections to check the AI engine.', 'run.error')],
      operationalCompleteness: requiredOutput ? 'incomplete' : 'not_assessed',
    };
  }

  return {
    outcome: 'failed',
    headline: statement(`${run.agentName} needs attention`, 'run.status'),
    summary: statement('The run stopped before it finished.', 'run.status'),
    problems: [statement('This run needs attention.', 'run.status')],
    suggestions: [statement('Review the problem before retrying.', 'run.status')],
    operationalCompleteness: requiredOutput ? 'incomplete' : 'not_assessed',
  };
}

function skippedMeaning(run: StoredRun): ReviewMeaning {
  const summary = run.code === 'lock_contention'
    ? 'Another run was already in progress. Nothing changed.'
    : run.code === 'already_completed_today'
      ? 'This assistant had already finished today. Nothing changed.'
      : 'The run was skipped. Nothing changed.';

  return {
    outcome: 'skipped',
    headline: statement(`${run.agentName} did not run`, 'run.status', 'run.code'),
    summary: statement(summary, 'run.code'),
    problems: [],
    suggestions: [],
    operationalCompleteness: 'not_assessed',
  };
}

function meaningFor(
  run: StoredRun,
  requiredOutput: RequiredOutput | undefined,
  pendingInteraction: PendingInteraction | undefined,
): ReviewMeaning {
  if (run.status === 'completed') return completedMeaning(run, requiredOutput);
  if (run.status === 'failed') return failedMeaning(run, requiredOutput);
  if (run.status === 'skipped') return skippedMeaning(run);
  if (run.status === 'running') {
    if (pendingInteraction) {
      return {
        outcome: 'waiting',
        headline: statement(
          `${run.agentName} is waiting for your response`,
          'interaction.status',
          'interaction.request',
        ),
        summary: statement(
          'The assistant needs your response before it can continue.',
          'interaction.status',
          'interaction.runId',
        ),
        problems: [],
        suggestions: [],
        operationalCompleteness: 'not_assessed',
      };
    }
    return {
      outcome: 'working',
      headline: statement(`${run.agentName} is working`, 'run.status'),
      summary: statement('Work is in progress.', 'run.status'),
      problems: [],
      suggestions: [],
      operationalCompleteness: 'not_assessed',
    };
  }

  return {
    outcome: 'unknown',
    headline: statement(`${run.agentName} has an unknown outcome`, 'run.status'),
    summary: statement('The recorded run state is not recognized.', 'run.status'),
    problems: [statement('The run state could not be interpreted.', 'run.status')],
    suggestions: [statement('Open Technical details for the recorded state.', 'run.status')],
    operationalCompleteness: 'not_assessed',
  };
}

function timelineFor(
  run: StoredRun,
  outcome: RunReview['outcome'],
  pendingInteraction: PendingInteraction | undefined,
): HumanTimelineEntry[] {
  const terminalTimestamp = (run.completedAt ?? run.startedAt).toISOString();
  const entries: HumanTimelineEntry[] = [{
    kind: 'started',
    label: statement('Started', 'run.startedAt'),
    occurredAt: run.startedAt.toISOString(),
  }];

  run.filesRead.forEach((path, index) => entries.push({
    kind: 'read',
    label: statement(`Read ${safeFileLabel(path)}`, `run.filesRead[${index}]`),
  }));
  run.filesWritten.forEach((path, index) => entries.push({
    kind: 'changed',
    label: statement(`Updated ${safeFileLabel(path)}`, `run.filesWritten[${index}]`),
  }));
  if (run.toolsUsed.length > 0) {
    entries.push({
      kind: 'connected',
      label: statement('Used a configured tool', 'run.toolsUsed'),
    });
  }

  if (pendingInteraction) {
    entries.push({
      kind: 'waiting',
      label: statement(
        'Waiting for your response',
        'interaction.status',
        'interaction.request',
      ),
      occurredAt: pendingInteraction.createdAt.toISOString(),
    });
  }

  if (run.status !== 'running') {
    const terminal = outcome === 'canceled'
      ? { kind: 'finished' as const, label: statement('Canceled', 'run.code') }
      : outcome === 'failed' || outcome === 'partial'
        ? { kind: 'problem' as const, label: statement('Stopped with a problem', 'run.status') }
        : { kind: 'finished' as const, label: statement('Finished', 'run.status') };
    entries.push({
      ...terminal,
      occurredAt: terminalTimestamp,
    });
  }
  return entries;
}

/** Translate durable local run evidence into a consumer-facing review. */
export function createRunReview(input: CreateRunReviewInput): RunReview {
  const { run, requiredOutput } = input;
  const pendingInteraction = currentPendingInteraction(
    run,
    input.pendingInteraction,
    input.now ?? new Date(),
  );
  const meaning = meaningFor(run, requiredOutput, pendingInteraction);
  const changes = run.filesWritten.map((path, index) => statement(
    `Updated ${safeFileLabel(path)}`,
    `run.filesWritten[${index}]`,
  ));
  const outputs = requiredOutput && run.status === 'completed'
    ? [statement(
      `${requiredOutput.label} is ready`,
      'agent.output.primary',
      'run.status',
    )]
    : [];

  return RunReviewSchema.parse({
    ...meaning,
    accomplishments: input.observedAccomplishments ?? [],
    changes,
    outputs,
    timeline: timelineFor(run, meaning.outcome, pendingInteraction),
    ...(pendingInteraction ? { waiting: waitingPresentation(pendingInteraction) } : {}),
    technicalDetailsReference: `/runs/${encodeURIComponent(run.runId)}`,
  });
}
